### DISCUSSION TITLE
nvim-dap extensibility

### DISCUSSION THREAD

**Discussion by LiadOz** [mfussenegger/nvim-dap#734] [General]
Hello,
I'm trying to implement a new type of breakpoints that allows triggering actions using hooks, this type of breakpoint will allow smarter debugging options, such as adding breakpoint A that can only start working after breakpoint B was hit.
Here is the repository https://github.com/LiadOz/meta-breakpoints.nvim
The main issues I have right now are that the nvim-dap is not extensible enough, for example I would like to place and remove breakpoints programmatically, right now I need to move the cursor to do that. Another thing that I would need is a way to change the behaviour that happens on breakpoint being hit. What are your thoughts about that?

**Comment by mfussenegger** [comment]
> for example I would like to place and remove breakpoints programmatically, right now I need to move the cursor to do that.

`require('dap.breakpoints').set` and `.toggle` both take a `bufnr` and `lnum` argument. No need to move the cursor.
But `dap.breakpoints` is currently considered internal and may change without deprecation notice.

> Another thing that I would need is a way to change the behaviour that happens on breakpoint being hit. 

Can you elaborate on the use-case? Overriding the whole event handler won't be supported as it sets some state that is important for other operations and extensions to work.
It also takes care of loading sources if they're not on the filesystem and so on. This would all break.

I could imagine providing an extension point or options to customize some of the jump behavior. This is a bit related to https://github.com/mfussenegger/nvim-dap/issues/676

But all of this needs problem descriptions and use-case scenarios. "Is not extensible enough" is not actionable.

**Reply by LiadOz** [reply]
> require('dap.breakpoints').set and .toggle both take a bufnr and lnum argument

This does not seem to work once the session has been initialized, from what I understand in order for the breakpoints to be sent to the server `session:set_breakpoints` should be called and the only places using this function are in dap.lua and in the initialization event.

> Another thing that I would need is a way to change the behavior that happens on breakpoint being hit.

I wasn't clear enough with this, the feature request here is to have option to disable jumping to breakpoint on hit (jump_to_frame) and also to have the abillity to get notified when breakpoints are hit. In my case when after breakpoint is hit I would want to set another breakpoint somewhere else. 

If you think these requests are valid I could create PR for them

**Reply by mfussenegger** [reply]
> This does not seem to work once the session has been initialized, from what I understand in order for the breakpoints to be sent to the server session:set_breakpoints should be called and the only places using this function are in dap.lua and in the initialization event.

Yes, the breakpoints need to be sent to the debug adapter to inform it about the changed breakpoints. But that could be done with something like this:

```lua
local breakpoints = require('dap.breakpoints')
local dap = require('dap')

local session = dap.session()
if session then
  session:set_breakpoints(breakpoints.get())
end
```

> the feature request here is to have option to disable jumping to breakpoint on hit (jump_to_frame)

Can you elaborate on why you'd want that?

> also to have the ability to get notified when breakpoints are hit. In my case when after breakpoint is hit I would want to set another breakpoint somewhere else.

You can subscribe to events, see `:help dap-extensions`

Dependent breakpoints is by the way something I'd like to add eventually to the core itself, but I'm still unsure of how it would be used from a UI perspective. Every time I have a use-case for them it seems to be less effort to just do it manually

**Reply by LiadOz** [reply]
As discussed I have created the following functions to set breakpoints at locations programmatically
```lua
local function setup_breakpoint_location(bp_location)
    bp_location = bp_location or {}
    local bufnr = bp_location.bufnr
    local lnum = bp_location.lnum
    if bufnr == nil and lnum == nil then
        bp_location.bufnr = vim.api.nvim_get_current_buf()
        bp_location.lnum = vim.api.nvim_win_get_cursor(0)[1]
    elseif not (bufnr ~= nil and lnum ~= nil) then
        error("Bad location passed, must use both bufnr and lnum")
    end
    return bp_location
end

local function toggle_dap_breakpoint(bp_opts, replace_old, bp_location)
    bp_opts = bp_opts or {}
    bp_location = setup_breakpoint_location(bp_location)
    bp_opts.replace = replace_old
    local bufnr = bp_location.bufnr
    local lnum = bp_location.lnum

    breakpoints.toggle(bp_opts, bufnr, lnum)
    local session = dap.session()
    if session then
        local bps = breakpoints.get(bufnr)
        session:set_breakpoints(bps)
    end
end
```
shouldn't this logic be implemented in nvim-dap as an api call?

> You can subscribe to events, see :help dap-extensions

I have not found a way to figure out that a breakpoint was hit with the current events, in the spec `event_stopped` could return the breakpoint that was hit, but not every debug adapter supports it, the only way for me to do that is to duplicate the logic inside event_stopped. The request here is to add a new event (not protocol event) that notifies a breakpoint was hit.

**Reply by mfussenegger** [reply]
> shouldn't this logic be implemented in nvim-dap as an api call?

Maybe, if there's a compelling argument to be made why it should be?

> the only way for me to do that is to duplicate the logic inside event_stopped

nvim-dap currently has no logic that tries to detect which breakpoint caused the stopping event. All it does is retrieve the frames and then jump to the top frame.

You could listen to the `stackTrace` response, or the and then access `current_frame` of the session to retrieve that info.
